### PR TITLE
BUGFIX: Allow DateTime for unserialization of form state

### LINKED ISSUE #126: DatePicker field in multi-page forms can't get unserialized
Using a DatePicker field in a multi-page form throws an exception, when you try to re-render the field with a value set.

E.g.

1. you set a Date on the first page
2. go further to the next page
3. step back to previous page

**Exception**
``` 
Could not convert target type “DateTime”: No converter found which can be used to convert from “__PHP_Incomplete_Class” or “object” to “DateTime”.
```

The unserialize of the form state causes this, because it is restricted to `FormState` class, so the `DateTime` object is ignored and handled as incomplete class.

This behavior was introduced by this changes:
https://github.com/neos/form/commit/c186327d9ada84e48fcee05b4828e4da1a8d1e60
https://github.com/neos/form/commit/ee593c2111515f6cc6ad6a77c13e234c1be5b8f2

Affected: Neos.Form >= 5.0

See also: https://discuss.neos.io/t/datetime-conversion-in-form/5007

### PR DESCRIPTION
**Pull request by dlubitz**
Fixes #126

### CODE REVIEW

File: Classes/Core/Runtime/FormRuntime.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
  * source code.
  */
 
+use DateTime;
```
Comment (albe) [inline Classes/Core/Runtime/FormRuntime.php:14]: Huh, that import is not needed, is it?

File: Classes/Core/Runtime/FormRuntime.php [hunk 2]
```diff
@@ -181,7 +182,7 @@ protected function initializeFormStateFromRequest()
             $this->formState = new FormState();
         } else {
             $serializedFormState = $this->hashService->validateAndStripHmac($serializedFormStateWithHmac);
-            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class]]);
+            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, DateTime::class]]);
```
Comment (bwaidelich) [inline Classes/Core/Runtime/FormRuntime.php:185]: Does this work as well:
```suggestion
            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTimeInterface::class]]);
```
?

File: Classes/Core/Runtime/FormRuntime.php (hunk 1 above)
Comment (dlubitz) [reply Classes/Core/Runtime/FormRuntime.php]: Ok. Done

File: Classes/Core/Runtime/FormRuntime.php (hunk 2 above)
Comment (dlubitz) [reply Classes/Core/Runtime/FormRuntime.php]: No. Unfortunately not.

File: Classes/Core/Runtime/FormRuntime.php [hunk 3]
```diff
@@ -181,7 +181,7 @@ protected function initializeFormStateFromRequest()
             $this->formState = new FormState();
         } else {
             $serializedFormState = $this->hashService->validateAndStripHmac($serializedFormStateWithHmac);
-            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class]]);
+            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTime::class]]);
```
Comment (bwaidelich) [inline Classes/Core/Runtime/FormRuntime.php:184]:
```suggestion
            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTime::class, \DateTimeImmutable::class]]);
```

**Comment by dlubitz** [comment]
Ready for merge and release? Or do you still have any objections @bwaidelich ?

**Comment by bwaidelich** [comment]
A test would be nice of course, but no general objections.

File: Classes/Core/Runtime/FormRuntime.php [hunk 4]
```diff
@@ -181,7 +181,7 @@ protected function initializeFormStateFromRequest()
             $this->formState = new FormState();
         } else {
             $serializedFormState = $this->hashService->validateAndStripHmac($serializedFormStateWithHmac);
-            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class]]);
+            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTime::class, \DateTimeImmutable::class]]);
```
Comment (kdambekalns) [inline Classes/Core/Runtime/FormRuntime.php:184]: Wouldn't that work, too?
```suggestion
            $this->formState = unserialize(base64_decode($serializedFormState), ['allowed_classes' => [FormState::class, \DateTimeInterface::class]]);
```
Answer: no, see https://3v4l.org/tdHfl

**Review by kdambekalns — APPROVED** [review]
I wonder if using `serialize()`/`unserialize()` to pass the form state is a good idea at all, but… 🤷‍♂️

**Comment by bwaidelich** [comment]
> I wonder if using serialize()/unserialize() to pass the form state is a good idea at all, but

Fair point, but what would you suggest instead?

**Comment by kdambekalns** [comment]
> > I wonder if using serialize()/unserialize() to pass the form state is a good idea at all, but
> 
> Fair point, but what would you suggest instead?

For the CR node properties we switched to JSON ages ago. But… 🤷‍♂️

**Comment by bwaidelich** [comment]
> For the CR node properties we switched to JSON ages ago

Yes, true that.. Would be a breaking change though because currently you can basically put any object into the form state

**Comment by bwaidelich** [comment]
> you can basically put any object into the form state

..OTOH you actually _can't_ as we see with this fix o.O

**Comment by kdambekalns** [comment]
Yes. When I added that restriction I (casually) checked `FormState` and assumed it only contains "harmless" things. Date objects might still count as harmless and that was just an oversight from my side.

But I did not realize one might actually have objects in a form that need to be stored in the `FormState` in a way that needs serialization…

**Pull request merged**